### PR TITLE
feat(wallet): add transactions_sort_by function

### LINKED ISSUE #794: Transaction Sorting and Querying
**Describe the enhancement**  
We'd like to be able to sort transactions at the DB layer and be able to paginate (and maybe in the future search) queries to the transactions.

**Use case**  
We want to display transactions to our customers in a sorted manner. We can do this by first querying all transactions and then sorting them after the fact, but this won't perform well once customers have a large number of transactions.

Currently we are thinking of sorting based on confirmation time for confirmed transactions and no sorting for unconfirmed transactions, but we may explore associating metadata with transactions (i.e. "initiation date" for when the transaction was initiated) and would want to be able to sort transactions based on that (this would help us sort unconfirmed transactions).

In the future, we may also allow customers to filter / search their transactions – ideally we could associate some sort of metadata here that we could filter / search by, but in a simpler use case could also imagine looking up date ranges (i.e. searching / grouping by confirmation time).

### PR DESCRIPTION
**Pull request by notmandatory**
### Description

Added new type alias `WalletTx` which represents a `CanonicalTx<'a, Arc<Transaction>, ConfirmationTimeHeightAnchor>` and new `Wallet::transactions_sort_by` that returns a `Vec<WalletTx>` sorted by the given compare function. 

### Notes to the reviewers

fixes bitcoindevkit/bdk#794 

### Changelog notice

* Add new type alias `WalletTx` which represents a `CanonicalTx<'a, Arc<Transaction>, ConfirmationTimeHeightAnchor>`.
* Add `Wallet::transactions_sort_by()` that returns a `Vec<WalletTx>` sorted by a given compare function. 

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

### CODE REVIEW

File: crates/wallet/src/wallet/mod.rs [hunk 1]
```diff
@@ -1136,15 +1137,33 @@ impl Wallet {
...
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// # use bdk_wallet::Wallet;
+    /// # use bdk_wallet::wallet::WalletTx;
+    /// # let changeset = Default::default();
+    /// # let mut wallet = Wallet::load_from_changeset(changeset)?;
+    /// // Transactions by chain position: first unconfirmed then descending by confirmed height.
+    /// let sorted_txs:Vec<WalletTx> = wallet.transactions_sort_by(|tx1, tx2| tx2.chain_position.cmp(&tx1.chain_position));
+    /// # Ok::<(), anyhow::Error>(())
```
Comment (storopoli) [inline crates/wallet/src/wallet/mod.rs:1122]: We are missing a closing ```
```suggestion
    /// # Ok::<(), anyhow::Error>(())
    /// ```
```

**Review by storopoli — COMMENTED** [review]
A minor comment.

File: crates/wallet/src/wallet/mod.rs [hunk 2]
```diff
@@ -1136,15 +1137,33 @@ impl Wallet {
...
+    pub fn transactions_sort_by<F>(&self, compare: F) -> Vec<WalletTx>
+    where
+        F: FnMut(&WalletTx, &WalletTx) -> Ordering,
+    {
+        let mut txs: Vec<WalletTx> = self.transactions().collect();
+        txs.sort_by(compare);
```
Comment (ValuedMammal) [inline crates/wallet/src/wallet/mod.rs]: perf: Consider using `sort_unstable_by` ?

File: crates/wallet/src/wallet/mod.rs (hunk 1 above)
Comment (notmandatory) [reply crates/wallet/src/wallet/mod.rs:1122]: oops! fixed.

File: crates/wallet/src/wallet/mod.rs (hunk 2 above)
Comment (notmandatory) [reply crates/wallet/src/wallet/mod.rs]: I didn't notice this, nice optimization.  Fixed.

**Review by ValuedMammal — APPROVED** [review]
ACK fb1599d50a38b37763a1246ba2c0de64587a150d

This looks good. I'm curious if users will also be interested in a list that includes transactions that may not be in the current active chain as was mentioned in https://github.com/bitcoindevkit/bdk/issues/794#issuecomment-2002656625

**Comment by notmandatory** [comment]
> I'm curious if users will also be interested in a list that includes transactions that may not be in the current active chain as was mentioned in [#794 (comment)](https://github.com/bitcoindevkit/bdk/issues/794#issuecomment-2002656625)

I like the idea of being able to show TX that are not in the valid chain due to RBF or reorg for a future PR. The idea with this one is to provide only the basic sort functionality for lang bindings to expose.

**Review by evanlinjin — APPROVED** [review]
ACK fb1599d50a38b37763a1246ba2c0de64587a150d

**Comment by notmandatory** [comment]
This only adds a new function and doesn't break any existing APIs so I'm moving this to the beta milestone so we can focus on breaking changes.

File: crates/wallet/src/wallet/mod.rs [hunk 3]
```diff
@@ -1037,13 +1041,10 @@ impl Wallet {
     /// ```
     ///
     /// [`Anchor`]: bdk_chain::Anchor
-    pub fn get_tx(
-        &self,
-        txid: Txid,
-    ) -> Option<CanonicalTx<'_, Arc<Transaction>, ConfirmationTimeHeightAnchor>> {
+    pub fn get_tx(&self, txid: Txid) -> Option<WalletTx> {
```
Comment (oleonardolima) [inline crates/wallet/src/wallet/mod.rs:1052]: @notmandatory I think the documentation needs an update now, referring to `WalletTx` instead, wdyt ?

**Review by oleonardolima — COMMENTED** [review]
Looks good! I just left a single comment about the documentation.

**Comment by oleonardolima** [comment]
Also, we'll still need to work on https://github.com/bitcoindevkit/bdk_wallet/issues/106 in the future, right ?

**Comment by storopoli** [comment]
> Also, we'll still need to work on bitcoindevkit/bdk_wallet#106 in the future, right ?

Yes because this is a simple `collect()` into a `Vec` and then sorting the vec in the Wallet API.
bitcoindevkit/bdk_wallet#106 is in the Chain API.

**Comment by evanlinjin** [comment]
@notmandatory I'm happy to merge this now tbh. It doesn't conflict or break anything.

**Comment by reez** [comment]
Just adding a +1 for merging this PR 🚀

File: crates/wallet/src/wallet/mod.rs (hunk 3 above)
Comment (notmandatory) [reply crates/wallet/src/wallet/mod.rs:1052]: Good catch, fixed.

**Comment by notmandatory** [comment]
Rebased and fixed docs.

**Comment by notmandatory** [comment]
Back by popular demand, merging this now instead of later.

**Comment by reez** [comment]
woohoo!

**Pull request merged**